### PR TITLE
o/snapstate: install components and snaps from file

### PR DESCRIPTION
**Pull request by andrewphelpsj**
This allows us to install snaps and components from files at the same time. Based on https://github.com/snapcore/snapd/pull/14092, first commit is 12afed26cc9e5e454233a7b838b0cbf26ad7dc54.

### CODE REVIEW

File: overlord/snapstate/target.go [hunk 1]
```diff
@@ -607,15 +682,39 @@ func (p *pathInstallGoal) toInstall(ctx context.Context, st *state.State, opts O
...
+		components: comps,
 	}
 
 	return []target{inst}, nil
 }
+
+func installableComponentsFromPaths(info *snap.Info, components map[*snap.ComponentSideInfo]string) ([]ComponentSetup, error) {
+	setups := make([]ComponentSetup, 0, len(components))
+	for csi, path := range components {
+		compInfo, _, err := backend.OpenComponentFile(path, info, csi)
```
Comment (pedronis) [inline overlord/snapstate/target.go:707]: shouldn't this do something more similar to validatedInfoFromPathAndSideInfo ?

**Review by pedronis — CHANGES_REQUESTED** [review]
I'm not sure if the code opening the component is doing enough validation, if I look at the similar path for the snap there's more validation going on

File: overlord/snapstate/target.go (hunk 1 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go]: Yes, resolved in 301c185f92b91b21466af283bbeba04436a40fb4.

File: overlord/snapstate/target.go [hunk 2]
```diff
@@ -677,15 +683,53 @@ func (p *pathInstallGoal) toInstall(ctx context.Context, st *state.State, opts O
...
+			CompPath:     path,
+			CompSideInfo: csi,
+			CompType:     compInfo.Type,
+		})
+	}
+
+	return setups, nil
+}
+
+func validatedComponentInfo(path string, csi *snap.ComponentSideInfo, si *snap.Info) (*snap.ComponentInfo, error) {
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/target.go:723]: nitpick: maybe order of arguments should be the same as for `backend.OpenComponentFile`

File: overlord/snapstate/snapstate_install_test.go [hunk 3]
```diff
@@ -6571,6 +6575,292 @@ func (s *snapmgrTestSuite) testInstallComponentsRunThrough(c *C, snapName, insta
...
+		tr.Set("core", "experimental.parallel-instances", true)
+		tr.Commit()
+	}
+
+	const snapID = "test-snap-id"
+	snapRevision := snap.R(11)
+	instanceName := snap.InstanceName(snapName, instanceKey)
+
+	compNameToType := func(name string) snap.ComponentType {
+		typ, ok := strings.CutSuffix(name, "-component")
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/snapstate_install_test.go:6658]: This is not in go 1.18 apparently.

File: overlord/snapstate/handlers_components.go [hunk 4]
```diff
@@ -291,6 +292,15 @@ func (m *SnapManager) doMountComponent(t *state.Task, _ *tomb.Tomb) (err error)
 	perfTimings.Save(st)
 	st.Unlock()
 
+	// if we're removing the snap file and we are mounting a component for the
+	// firs time, then we know that the component also must be coming from an
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/handlers_components.go:296]: typo firs -> first

**Review by alfonsosanchezbeato — COMMENTED** [review]
Some minor comments

File: overlord/snapstate/handlers_components.go [hunk 5]
```diff
@@ -291,6 +292,15 @@ func (m *SnapManager) doMountComponent(t *state.Task, _ *tomb.Tomb) (err error)
...
+	// emphemeral file. in that case, remove it.
+	if snapsup.RemoveSnapPath {
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/handlers_components.go:298]: Maybe we should add the flags from `componentInstallFlags` to `ComponentSetup` and check that instead?

Comment (andrewphelpsj) [reply overlord/snapstate/handlers_components.go:298]: Done in fef69533116301a41398f18f824541c18173eddd.

File: overlord/snapstate/snapmgr.go [hunk 6]
```diff
@@ -202,6 +202,9 @@ type ComponentSetup struct {
 	// DownloadInfo contains information about how to download this component.
 	// Will be nil if the component should be sourced from a local file.
 	DownloadInfo *snap.DownloadInfo `json:"download-info,omitempty"`
+	// Flags is a set of flags that control the behavior of the component
+	// installation/update.
+	Flags componentInstallFlags `json:"flags"`
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:207]: componentInstallFlags itself doesn't seem to have json tags, also should flags here be embedded like they are in SnapSetup?

File: overlord/snapstate/handlers_components.go (hunk 5 above)
Comment (pedronis) [reply overlord/snapstate/handlers_components.go:298]: but this is code still the same? what am I missing

File: overlord/snapstate/target.go [hunk 7]
```diff
@@ -677,15 +704,53 @@ func (p *pathInstallGoal) toInstall(ctx context.Context, st *state.State, opts O
...
+		components: comps,
 	}
 
 	return []target{inst}, nil
 }
+
+func componentSetupsFromPaths(snapInfo *snap.Info, components map[*snap.ComponentSideInfo]string) ([]ComponentSetup, error) {
+	setups := make([]ComponentSetup, 0, len(components))
+	for csi, path := range components {
+		compInfo, err := validatedComponentInfo(path, snapInfo, csi)
```
Comment (pedronis) [inline overlord/snapstate/target.go:729]: thank you, wondering did we miss any validation on the store flows?

File: overlord/snapstate/target.go [hunk 8]
```diff
@@ -677,15 +704,53 @@ func (p *pathInstallGoal) toInstall(ctx context.Context, st *state.State, opts O
...
+			CompType:     compInfo.Type,
+		})
+	}
+
+	return setups, nil
+}
+
+func validatedComponentInfo(path string, si *snap.Info, csi *snap.ComponentSideInfo) (*snap.ComponentInfo, error) {
+	if err := csi.Component.Validate(); err != nil {
+		return nil, err
```
Comment (pedronis) [inline overlord/snapstate/target.go:746]: this path doesn't seem to fail in tests?  also where do we check that the snap name and snap name in the component match?  is something we do later in the tasks? maybe we should do that early as well?

**Review by pedronis — COMMENTED** [review]
thank you for the changes, some further comments

File: overlord/snapstate/handlers_components.go (hunk 5 above)
Comment (andrewphelpsj) [reply overlord/snapstate/handlers_components.go:298]: Sorry, missed that in the commit.

File: overlord/snapstate/target.go (hunk 7 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:729]: Most of the validation happens in [componentSetupFromResource](https://github.com/snapcore/snapd/blob/master/overlord/snapstate/target.go?plain=1#L331). We check that the component from the resource exists in the `snap.yaml` that we get back from the store, and we check that the resource's type matches the type that is defined in the `snap.yaml`. I think that is really all we can do at that point?

And then there is the stuff that happens in `doValidateComponent` with assertions, of course.

File: overlord/snapstate/target.go (hunk 8 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:746]: That error is covered by TestInstallComponentsFromPathInvalidComponentName.

Comment (andrewphelpsj) [reply overlord/snapstate/target.go:746]: We check that the snap name and the snap name in the component match inside of [`snap.ReadComponentInfoFromContainer`](https://github.com/snapcore/snapd/blob/master/snap/component.go?plain=1#L228), which is called by `OpenComponentFile`.

**Review by pedronis — APPROVED** [review]
thank you

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks for the changes

**Pull request merged**